### ISSUE TITLE
Request for Additional Information and Code Clarification

### ISSUE THREAD

**Issue by KANGX99** [duchengbin8/Stable_Diffusion_is_Unstable#3]
Hi @duchengbin8,
Firstly, I would like to extend my sincere appreciation for your outstanding work. Your project has provided us with numerous valuable insights and has been incredibly enlightening.

I have a few questions and requests for further clarification:
1. Could you kindly **provide additional code snippets** that demonstrate how perturbations can be incorporated by replacing an existing word? Additionally, for attacking long texts, does this involve making modifications within the text rather than simply appending words?
2. **Hyperparameter 'lam_perp'**:
   - In your code, the 'lam_perp' parameter is set to 0. Could you please explain the rationale behind this choice? 
   - Is 'lam_perp' intended to represent the weight of the fluency constraint?
3. **Training Loss Convergence**:
   - I observed that `total_loss = adv_loss + ref_loss + perp_loss` does not converge during training. Below is an excerpt from our training log for reference:
     ```
     Iteration 1: loss = 22.2543, adv_loss = 25.9688, ref_loss = -3.7144, perp_loss = 0.0000, entropy=18.3518,           
     time=4.35
     Iteration 11: loss = 26.2216, adv_loss = 30.1562, ref_loss = -3.9346, perp_loss = 0.0000, entropy=13.5757, 
     time=44.63
     Iteration 21: loss = 27.0919, adv_loss = 31.0312, ref_loss = -3.9393, perp_loss = 0.0000, entropy=22.4976, 
     time=85.22     
     .....
     Iteration 91: loss = 22.5275, adv_loss = 26.4375, ref_loss = -3.9100, perp_loss = 0.0000, entropy=49.6613, 
     time=367.21
     ```

**Comment by duchengbin8** [comment]
1. In line 128 of `whitebox_attack.py`, we have the option to pad the text to a certain length. You can choose not to pad, in which case only existing words will be replaced. If you do choose to pad, the algorithm will automatically modify the tokens used for padding. However, in most cases, due to the constraints imposed by text similarity, not all padding tokens will be altered.

2.1 The values of the hyperparameters lam_sim and lam_perp used in the paper are 0.5 and 1, respectively. These values can be adjusted at will. They just represent a trade-off between attack success rate and text similarity. 

2.2 Yes, lam_perp is represent the weight of the fluency constraint.

3. When the input text changes, even if the random seed remains the same, the content of the images generated by the original stable diffusion model will inevitably change. This can lead to fluctuations in loss. However, experimental results show that despite this loss fluctuation, we can still find those adversarial prompts.

**Comment by KANGX99** [comment]
Hi, @duchengbin8, Thank you for the detailed response. Your help is greatly appreciated! I have a few more questions:

2. I observed that simply modifying line 128 of `whitebox_attack.py` does not directly enable modifications within the text. This is due to the presence of the `forbidden` list in lines 181-188, which prevents the matrix from being updated (lines 272-273). Would it be correct to set this `forbidden` list to empty in order to achieve the desired functionality?
3. Additionally, according to **Algorithm 1** in the paper, should the mask be applied to the position corresponding to the `classname`? I would greatly appreciate it if you could provide specific code snippets for this, as I am particularly interested in this work.

Once again, I commend you for your excellent work and I am looking forward to your response. Thank you for your time and assistance.

**Comment by duchengbin8** [comment]
2. In our paper, the forbidden mask is used to prevent the update of tokens related to category names. For example, if we want to modify based on "a photo of a tench," without forbidding the update of tokens related to "tench," the algorithm might replace "tench" with other words, such as "a photo of a cat". This is not considered an attack prompt. If you want to make direct modifications based on the prompt, simply remove the padding at line 128. If you want to extend the length of the prompt, you can change `model_max_length` at line 104. After using padding, for example, "a photo of a tench <‘pad’><‘pad’>", the algorithm will update the gradient at the <‘pad’>, After decoding, the unchanged <‘pad’> tokens will be automatically removed(line 293).

3. Lines 185-188 contain the code for setting the forbidden mask.

**Comment by KANGX99** [comment]
Yes, I understand what you mean. 
Taking "**_a photo of a goldfish_**" for example, I want to make direct modifications based on the prompt
> input_ids:  [49406, 320, 1125, 539, 320, 40293, 49407]
> unchange_ids:  [49406, 40293, 49407]
> forbidden:  [ True  False False False False True True]

Is this what you mean?

**Comment by duchengbin8** [comment]
yes

**Comment by KANGX99** [comment]
Thank you for your detailed answer, your work is so elegant!🉑
